### PR TITLE
Add tests for 'form-associated custom elements'

### PR DESCRIPTION
**Pull request by tkent-google**
This is for https://github.com/w3c/webcomponents/issues/187

Specification PR: https://github.com/whatwg/html/pull/4383

### CODE REVIEW

File: custom-elements/form-associated/ElementInternals-setFormValue.html [hunk 1]
```diff
@@ -0,0 +1,120 @@
...
+      '<my-control name=name-ce1></my-control>' +
+      '</form>' +
+      '<iframe name="if1"></iframe>';
+  $('my-control').value = 'value-ce1';
+  $('my-control').setValues([]);
+  return submitPromise(t).then(query => {
+    assert_equals(query, '?name-pd1=value-pd1');
+  });
+}, 'setFormValue with an empty FormData should submit nothing');
+</script>
```
Comment (domenic) [inline custom-elements/form-associated/ElementInternals-setFormValue.html]: We should test the different overloads:

- Test File (including that filename is preserved, i.e. it isn't downcast into a Blob)
- Test USVString lone surrogate conversion (idlharness is probably not smart enough)

I don't think there is any way to test the second argument and test formStateRestoreCallback reliably, but maybe you can think of one... I guess that would be a separate file, if we do test it.

We may want to just file an issue for testing formStateRestoreCallback in general, and the WPT infra team might be interested in helping us add infra for autofill WPTs and similar.

File: custom-elements/form-associated/ElementInternals-validation.html [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+  isInvalid(form);
+  isInvalid(fieldset);
+
+  control.i.setValidity({});
+  isValid(control);
+  isValid(form);
+  isValid(fieldset);
+}, ':valid :invalid for FORM and FIELDSET');
+</script>
+</body>
```
Comment (domenic) [inline custom-elements/form-associated/ElementInternals-validation.html]: We should test the shadow-including descendant requirement on the anchor element

**Review by domenic — COMMENTED** [review]
Sorry I missed this the first time through. Overall looking great.

File: custom-elements/form-associated/form-reset-callback.html [hunk 3]
```diff
@@ -0,0 +1,53 @@
...
+  output.addEventListener('DOMSubtreeModified', () => {
+    assert_false(custom.resetCalled, 'formResetCallback should not be ' +
+                 'called before built-in control\'s reset');
+  });
+  form.reset();
+  assert_true(custom.resetCalled);
+}, 'form.reset(): formResetCallback is called after reset of the last ' +
+   'built-in form control and before the next statement.');
+
+let t = async_test('Clicking a reset button invokes formResetCallback in a ' +
```
[the remark this answers is not in the document]
Comment (tkent-google) [reply custom-elements/form-associated/form-reset-callback.html]: Done.  This is now a promise_test().

File: custom-elements/form-associated/form-reset-callback.html [hunk 4]
```diff
@@ -0,0 +1,53 @@
...
+      '<form><my-control></my-control><input type=reset></form>');
+  let form = document.body.lastChild;
+  let custom = form.firstChild;
+  let resetButton = form.lastChild;
+  assert_false(custom.resetCalled);
+  resetButton.click();
+  assert_false(custom.resetCalled);
+  window.queueMicrotask(t.step_func_done(() => {
+    assert_true(custom.resetCalled);
+  }));
```
[the remark this answers is not in the document]
Comment (tkent-google) [reply custom-elements/form-associated/form-reset-callback.html]: Done.

File: custom-elements/form-associated/ElementInternals-setFormValue.html (hunk 1 above)
Comment (tkent-google) [reply custom-elements/form-associated/ElementInternals-setFormValue.html]: Added tests for USVString and File.

I don't think WPT can have tests for formStateRestoreCallback.  Its behavior can not be defined by standards.

**Review by domenic — APPROVED** [review]

**Comment by domenic** [comment]
r

**Pull request merged**